### ISSUE TITLE
[libeditor] Various bug fixes

### ISSUE THREAD

**Issue by g4jc** [MoonchildProductions/UXP#1561]
During my time poking around for `dom_work`, I discovered the following bug fixes that are partial blockers as well as bug fixes and improvements to libeditor. Since they've already applied cleanly I am listing them below for possible uplift.

- [x] [Bug 1316302](https://bugzilla.mozilla.org/show_bug.cgi?id=1316302) Backspace key does nothing when collapsed selection is in a block and its previous child block has a non-empty text node at rightmost 
- [x] [Bug 1318570](https://bugzilla.mozilla.org/show_bug.cgi?id=1318570) Clean up GetTextNode into TextEditRules
- [x] [Bug 564411](https://bugzilla.mozilla.org/show_bug.cgi?id=564411) Move all methods/attributes of nsIEditorIMESupport to nsIEditor
- [x] [Bug 1323138](https://bugzilla.mozilla.org/show_bug.cgi?id=1323138) CSSEditUtils should use atom for CSS property if possible.
- [x] [Bug 1324996](https://bugzilla.mozilla.org/show_bug.cgi?id=1324996) Add nsIAtom version of SetAttribute/RemoveAttribute/CloneAttirubte to Editor 
- [x] [Bug 1325300](https://bugzilla.mozilla.org/show_bug.cgi?id=1325300) Use explicitly static method call for GetStartNodeAndOffset and GetEndNodeAndOffset.
- [x] [Bug 1342197](https://bugzilla.mozilla.org/show_bug.cgi?id=1342197) Speed up GetSelectionRange in text controls 
- [x] [Bug 1345763](https://bugzilla.mozilla.org/show_bug.cgi?id=1345763) Create AsTextEditor() and AsHTMLEditor()
- [x] [Bug 1318570](https://bugzilla.mozilla.org/show_bug.cgi?id=1343037) Move various selection-management stuff out of nsTextControlFrame and into the editor state 
- [x] [Bug 1343275](https://bugzilla.mozilla.org/show_bug.cgi?id=1343275) Make the text control editor state's GetSelectionRange codepath a bit faster when going via the selection controller 
- [x] [Bug 1345606](https://bugzilla.mozilla.org/show_bug.cgi?id=1345606) setRangeText should mark the text control as dirty
- [x] [Bug 1345237](https://bugzilla.mozilla.org/show_bug.cgi?id=1345237) Propagate uint32_t deeper into the editor state and text control frame code.
- [x] [Bug 1357206](https://bugzilla.mozilla.org/show_bug.cgi?id=1357206) Don't move selection to end when .value is set to its existing value 
- [x] [Bug 1364814](https://bugzilla.mozilla.org/show_bug.cgi?id=1364814) Use RAII class to set and restore editor flags.
- [x] [Bug 1358025](https://bugzilla.mozilla.org/show_bug.cgi?id=1358025) Don't create transaction when script sets input.value 
- [x] [Bug 1349178](https://bugzilla.mozilla.org/show_bug.cgi?id=1349178) Make nsIEditor.transactionManager readonly
- [x] [Bug 1350772](https://bugzilla.mozilla.org/show_bug.cgi?id=1318570) Crash Fix
- [x] [Bug 1431336](https://bugzilla.mozilla.org/show_bug.cgi?id=1431336) Crash Fix

**Comment by mattatobin** [comment]
They won't be uplifted unless you distinguish between what is required for `WebComponents` and what makes sense.

**Comment by g4jc** [comment]
>  what is required for WebComponents
None of this is technically _required_ for `WeCcomponents `which is why I made a separate issue for it. It just touches a number of files that we will be touching again later. Notably Bug 1323138 fixes a crash in CSSEditUtils  which we do want to have stable for further work there.

> When were they done?
They're listed in chronological order of commit date.

> And what commits are these?
 Bug  1316302 fixes an issue where backspace does not work, Bug 1323138 fixes a crash in CSSEditUtils (the same files are touched later on for Webcomponents), Bug 1345606 and 1345237 are web specification fixes. The rest are code cleanup, optimize, and sec enhancement (including a memory leak in 1349178). I have not yet tested these changes against `mailnews` on Windows which will need to be done prior to any merge.

I have confirmed at minimum `GetSelectionRange` is working well after speedup in IW, test passing per http://jsfiddle.net/5vd8pxct/

**Comment by mattatobin** [comment]
Well you need to actually include the commits that should be uplifted and an estimation of the risks involved.

**Comment by wolfbeast** [comment]
All of these checkboxes are checked. So... where's the code that belongs to these checked items? Which commits are we looking at, if any? I can't do anything with this in terms of code adoption.

**Comment by g4jc** [comment]
Patches landed in a local branch here, in the future I'll try and avoid having checked boxes since it caused some confusion. I checked them off simply because they applied cleanly and we're already in my libeditor branch.

https://github.com/g4jc/UXP/tree/libeditor_patch

As per request I'm going to break this up later into separate issues since we can't land this as a meta.

**Comment by wolfbeast** [comment]
Was all of this landed?

**Comment by mattatobin** [comment]
No.

**Comment by wolfbeast** [comment]
*sigh* so what still needs to be done of the list that is all checked off?

**Comment by mattatobin** [comment]
No idea, likely proper cleanup and landing of commits.. @g4jc IF THIS IS DONE, please give me your branch and I will land them.

**Comment by g4jc** [comment]
Only 1316302 was landed (updated post to reflect), the rest are still in my branch if we decide we still need them. Currently this isn't blocking anything so is low priority.

**Comment by wolfbeast** [comment]
I guess I'll go through the list then and see what we want, what we don't want, and if there's anything missing.

**Comment by wolfbeast** [comment]
1. [Y] Bug 1318570 Clean up GetTextNode into TextEditRules
2. [?] Bug 564411 Move all methods/attributes of nsIEditorIMESupport to nsIEditor
=> if we take this, we also need to take the follow-up 1325281 and that requires comm to adjust following the changes in 1325282
3. [Y] Bug 1323138 CSSEditUtils should use atom for CSS property if possible.
=> Free perf win and fixes assertion bug 1274050
4. [Y] Bug 1324996 Add nsIAtom version of SetAttribute/RemoveAttribute/CloneAttirubte to Editor
=> Free perf win but needs 16 to fix regression crash
5. [Y] Bug 1325300 Use explicitly static method call for GetStartNodeAndOffset and GetEndNodeAndOffset.
=> Removes a level of indirection
6. [Y] Bug 1342197 Speed up GetSelectionRange in text controls
=> Is a significant increase in perf although unsure if it matters much in real world scenarios. May as well take it if the work has already been done.
7. [Y] Bug 1345763 Create AsTextEditor() and AsHTMLEditor()
=> Useful conversion methods
8. [Y] Bug 1343037 Move various selection-management stuff out of nsTextControlFrame and into the editor state
=> Follow-up to 6 to address related perf bottlenecks
9. [Y] Bug 1343275 Make the text control editor state's GetSelectionRange codepath a bit faster when going via the selection controller
=> Follow-up to 6 to address related perf bottlenecks
10. [Y] Bug 1345606 setRangeText should mark the text control as dirty
=> Follow-up to 6 to address behavior change, re-aligning it with the current spec
11. [Y] Bug 1345237 Propagate uint32_t deeper into the editor state and text control frame code.
=> Aligns types used with purpose (code correctness)
12. [Y] Bug 1357206 Don't move selection to end when .value is set to its existing value
=> Spec changed, we need to follow this with the change in this bug
13. [N] Bug 1364814 Use RAII class to set and restore editor flags.
=> I discussed this in e-mail with @g4jc and don't understand what it's still doing here. It's nonsense to use an RAII to set/restore flags.
14. [Y] Bug 1358025 Don't create transaction when script sets input.value
=> Bug described best in the dupe 1283912. It makes no sense to allow Ctrl+Z undo to undo something that isn't user input.
15. [Y] Bug 1349178 Make nsIEditor.transactionManager readonly
=> removes a footgun
16. [Y] Bug 1350772 Crash Fix for regression from 4.
17. [Y] Bug 1431336 Crash Fix for regression from 8.
 
**Bug groups for issues:**
Issue 1: 1.
Issue 2?: 2, 1325281 and 1325282. 
Issue 3: 3, 4, and 16.
Issue 4: 5.
Issue 5: 6, 8, 9, 10 and 17.
Issue 6: 7.
Issue 7: 11.
Issue 8: 12.
Issue 9: 14.
Issue 10: 15.

**Comment by g4jc** [comment]
Thanks for reviewing.

I think we can skip 564411, the two follow ups mention removing addons prior to removing the methods.
Making a PR for issue 1.
